File: benchmarks/bench_stack.py
#!/usr/bin/env python3
"""bench_stack.py — Python equivalent of the Aria stack benchmarks."""
import time

N = 10_000_000

# Warm up
stack = []
s = 0
for w in range(1000):
    stack.append(w * 3 + 1)
    stack.append(w * 7 + 2)
    wb = stack.pop()
    wa = stack.pop()
    s += wa + wb

stack.clear()
s = 0

t_start = time.monotonic_ns()

for i in range(N):
    stack.append(i * 3 + 1)
    stack.append(i * 7 + 2)
    b = stack.pop()
    a = stack.pop()
    s += b + a

t_end = time.monotonic_ns()

elapsed_ns = t_end - t_start
elapsed_ms = elapsed_ns // 1_000_000

print("=== PYTHON LIST-AS-STACK ===")
print(f"  iterations: {N}")
print(f"  sum: {s}")
print(f"  elapsed_ms: {elapsed_ms}")
print(f"  elapsed_ns: {elapsed_ns}")

File: benchmarks/bench_stack.js
#!/usr/bin/env node
// bench_stack.js — Node.js equivalent of the Aria stack benchmarks.
"use strict";

const N = 10_000_000;

// Warm up
const stack = [];
let s = 0;
for (let w = 0; w < 1000; w++) {
    stack.push(w * 3 + 1);
    stack.push(w * 7 + 2);
    const wb = stack.pop();
    const wa = stack.pop();
    s += wa + wb;
}
stack.length = 0;
s = 0;

const t_start = process.hrtime.bigint();

for (let i = 0; i < N; i++) {
    stack.push(i * 3 + 1);
    stack.push(i * 7 + 2);
    const b = stack.pop();
    const a = stack.pop();
    s += b + a;
}

const t_end = process.hrtime.bigint();
const elapsed_ns = Number(t_end - t_start);
const elapsed_ms = Math.floor(elapsed_ns / 1_000_000);

console.log("=== NODE.JS ARRAY-AS-STACK ===");
console.log(`  iterations: ${N}`);
console.log(`  sum: ${s}`);
console.log(`  elapsed_ms: ${elapsed_ms}`);
console.log(`  elapsed_ns: ${elapsed_ns}`);
